### ISSUE TITLE
Galvanostatic mode

### ISSUE THREAD

**Issue by j-fu** [WIAS-PDELib/VoronoiFVM.jl#131]
Continue discussion from #119

**Comment by j-fu** [comment]
Just tried out your MWE, it works. Just a hint: with recent SciML versions, it is sufficient to use OrdinaryDiffEqBDF instead of the whole of DifferentialEquations. 

Will take some time to think about galvanostatic (next week is very busy...)

**Comment by JanSuntajs** [comment]
Thank you; for now, I probably can make it work just by assigning another species for the voltage and then look for solutions satisfying Laplace equation with homogenous Neumann b. c. at the center and a properly chosen Dirichlet b. c. at the external boundary. This is of course a bit redundant since it doubles the storage requirements, but probably still feasible in 1D.

Best,
Jan

**Comment by JanSuntajs** [comment]
@j-fu on a somehow related note to the above discussion, I am now trying to expand the above functionality. I am wondering, whether it is possible access global solution vector from within the standard functions such as ```storage!(...), reaction!(...), flux!(...)```. Namely, on a given node, is it possible to access degrees of freedom that are not defined on a given node or on nodes directly adjacent to it? I know this can be done by the means of the generic operators, but I am not yet skilled enough with them, particularly regarding the manual sparsity detection. I am also unsure about how to correctly write down the generic operator, are the changes of coordinate systems (for instance, cartesian vs. curvilinear coordinates) handled automatically or should I do them on my own? 

Thank you and best regards,

Jan

**Comment by j-fu** [comment]
Conceptually, there is no way to access the global solution vector from the callbacks. However you could trick around this by having a SystemState (since v2.0) as part of the userdata. The drawback is that there will be no differentiation against the corresponding DOFs, in the best case resulting in detoriation of Newton convergence to linear. The generic operator has been created to handle this situation in a consistent manner. 

I know it is not well documented, so if you have a minimal example covering the situation you need  I would take the occasion to update the docs. 
When accessing node volumes and edge coefficients, these are precalculated with regard to the coordinate system. 

Also, there is the possibility to detect the sparsity pattern automatically. This step may take some time, but is needed only once in a simulation.

**Comment by JanSuntajs** [comment]
Thank you, I will try to come up with a useful example; for now, it puzzles me a bit that I do not get the same solutions if I just try to rewrite the reaction term as a generic operator. At this point, I do not yet try to access dofs in non-adjacent nodes, I just try to come up with an equivalent formulation in terms of a generic operator, but to no avail. I will dig into this some more and then provide a minimal example.

Best,
Jan

**Comment by JanSuntajs** [comment]
@j-fu, in the attached jupyter notebook, I have implemented a MWE that displays the difference between implementing the same equation using the ```reaction!(f, u, node, data)``` function and via ```generic_operator(...)```. In the latter case, the solver fails to reproduce the results of the former one. That is even before I tried calculating with generic operators (also included in the notebook) which would consider the surface concentration that I mentioned in above posts.

I am also attaching the manifest and project files.
[double_layer.zip](https://github.com/user-attachments/files/17700470/double_layer.zip)

Please let me know if everything works as expected and whether anything should be changed.

Regards,
Jan

**Comment by j-fu** [comment]
Ok, I tried to run it, but I get "UndefVarError: `data_dl` not defined" in the last cell.

Generally it would be more convenient for me to share code with a git repo, and I also have a slight preference for pluto notebooks. 
But in fact I did not use jupyter for quite some time since I switched to pluto, and I just realized that installing the jupyter stack via conda is quite smooth nowadays, so I might be ok with it.

**Comment by JanSuntajs** [comment]
I have shared with you (for now a private) git repo, in which the selected notebook can be found under ```notebooks/05_double_layer.ipynb``` and the mentioned error was corrected. 

For convenience, I also include the zip archive here:
[mwe.zip](https://github.com/user-attachments/files/17704330/mwe.zip)

(it is beneficial to install IJulia package to the global Julia environment if working with Jupyter notebooks). Since I have not used Pluto before, I can also prepare some scripts in case the above does not work).

**Comment by jpthiele** [comment]
@JanSuntajs If I understood your comments in #189 correctly this issue here is resolved, right?
If not, feel free to reopen!

**Comment by JanSuntajs** [comment]
> [@JanSuntajs](https://github.com/JanSuntajs) If I understood your comments in [#189](https://github.com/WIAS-PDELib/VoronoiFVM.jl/issues/189) correctly this issue here is resolved, right? If not, feel free to reopen!

True, the bare galvanostatic problem now works with boundary species as well. In addition, I learned there is no need to manually handle the algebraic equation for the boundary condition by means of root finding as any appropriate solver will do this on its own. I can prepare an illustratory MWE if needed to showcase the workflow.

The problem from issue #189 of course still stands.
